### DISCUSSION TITLE
Simpler way to implement a Windows SCM service.

### DISCUSSION THREAD

**Discussion by The-Futurist** [dotnet/runtime#101511] [Q&A] [answered]
I understand that the Windows Service model and SCM, are platform specific but is there really no easier way to code a service now, other than this [rather involved](https://learn.microsoft.com/en-us/dotnet/core/extensions/windows-service) and convoluted "recipe"?

Who has the time to "rewrite" this and that class and all the other steps. I've built Windows services many times in both C and C# (.Net F) and think we need a simpler approach for doing so under .Net.

There isn't even a template for such a project, so I'm asking what would it take to create a .Net version of the older `ServiceBase` and other classes? Having an _identical_ set of classes to what was used in .Net F would be so helpful.

**Comment by tacosontitan** [comment]
As someone who actively works with Windows services in both .NET Framework and modern .NET, I completely disagree with the idea of having an identical set of classes for that. I prefer the modern worker service approach much more than the classic `ServiceBase` approach.

There *is* a template for worker services and this template can easily be extended to add support for a Windows service, and there isn't a ton of rewriting needed to facilitate it.

https://learn.microsoft.com/en-us/dotnet/core/extensions/windows-service

**Reply by The-Futurist** [reply]
Hi, @tacosontitan  I'd like to simply rebuild an existing service app, with a plug-in compatible set of service classes with the same interfaces as we had in .Net F. I would prefer to avoid changing the design at all. I'm sure you are right and that the more "modern" design is fine, but I'm seeking to port a complex service and - like any sensible developer - I'm trying to do as little as possible!

In that article, where are the various optional events that a service can handle? like shutdown, start, pause, continue, [power fail](https://learn.microsoft.com/en-us/dotnet/api/system.serviceprocess.powerbroadcaststatus?view=dotnet-plat-ext-8.0) and so on?

I wonder if a nuget package could be written that mimics the older .Net F service classes...

**Comment by The-Futurist** [comment]
Hmm, seems [there is a .Net 8 version](https://learn.microsoft.com/en-us/dotnet/api/system.serviceprocess?view=dotnet-plat-ext-8.0) of all this, how did I miss this?

**Comment by The-Futurist** [comment, marked as answer]
Yup, seems exactly what I was seeking, amazing!

![image](https://github.com/dotnet/runtime/assets/12262952/c3f766cb-1580-4c70-8aa3-b4cb30b354bc)
